### PR TITLE
Fix TypeCastExpression

### PR DESCRIPTION
**Pull request by czpmango**
This PR fixes TypeCastExpression behavior(refer to: [openCypher scalar functions](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-toboolean )) and add functions related with type conversion.

Depended by [https://github.com/vesoft-inc/nebula-graph/pull/746](https://github.com/vesoft-inc/nebula-graph/pull/746) and [https://github.com/vesoft-inc/nebula-storage/pull/352](https://github.com/vesoft-inc/nebula-storage/pull/352)

Fixes [https://github.com/vesoft-inc/nebula-graph/issues/677](https://github.com/vesoft-inc/nebula-graph/issues/677) and [https://github.com/vesoft-inc/nebula-graph/issues/706](https://github.com/vesoft-inc/nebula-graph/issues/706)

---

Note:
```
Value::toBool() 行为有如下变动:
 - 除 null/bool/string 之外的类型视为非法类型，返回 kNullBadType
 - bool 类型原样返回
 - 字符串 "true"/"FaLse" 等会返回转换后的 bool 值
 - null 和其他字符串会返回 null
 - empty 类型行为暂未定义，目前返回 null
```

### CODE REVIEW

File: src/common/datatypes/Value.cpp [hunk 1]
```diff
@@ -1550,79 +1550,72 @@ std::string Value::toString() const {
...
         }
         case Value::Type::STRING: {
-            return std::make_pair(!getStr().empty(), true);
-        }
-        case Value::Type::DATE: {
-            return std::make_pair(getDate().toInt() != 0, true);
+            if (toString().compare("true") == 0) {
+                return true;
+            }
+            if (toString().compare("false") == 0) {
```
Comment (yixinglu) [inline src/common/datatypes/Value.cpp:1566]: you'd better to assign the result of `toString()` to a variable for avoiding this function to be called multiple times.

Comment (czpmango) [reply src/common/datatypes/Value.cpp]: Fixed, thx.

File: src/common/datatypes/Value.cpp [hunk 2]
```diff
@@ -1550,79 +1550,74 @@ std::string Value::toString() const {
...
-            return std::make_pair(std::abs(getFloat()) > kEpsilon, true);
+            return *this;
         }
         case Value::Type::STRING: {
-            return std::make_pair(!getStr().empty(), true);
-        }
-        case Value::Type::DATE: {
-            return std::make_pair(getDate().toInt() != 0, true);
+            auto str = toString();
+            std::transform(str.begin(), str.end(), str.begin(), ::tolower);
```
Comment (jievince) [inline src/common/datatypes/Value.cpp:1564]: why not directly call std::lower(toString())?

Comment (czpmango) [reply src/common/datatypes/Value.cpp:1564]: std::tolower?

Comment (jievince) [reply src/common/datatypes/Value.cpp:1564]: I'am sorry. Seems std::string doesn't has tolower()

File: src/common/datatypes/Value.cpp [hunk 3]
```diff
@@ -1550,79 +1550,74 @@ std::string Value::toString() const {
     LOG(FATAL) << "Unknown value type " << static_cast<int>(type_);
 }
 
-std::pair<bool, bool> Value::toBool() {
+Value Value::toBool() {
```
Comment (panda-sheep) [inline src/common/datatypes/Value.cpp:1553]: please convert to bool in this function

**Review by panda-sheep — CHANGES_REQUESTED** [review]
please convert Value::toBool  to bool

File: src/common/datatypes/Value.cpp (hunk 3 above)
Comment (czpmango) [reply src/common/datatypes/Value.cpp:1553]: You could get the bool type return value explicitly in the caller interface.

Comment (panda-sheep) [reply src/common/datatypes/Value.cpp:1553]: I don’t think so, I think you should check the original function implementation

**Comment by critical27** [comment]
> - 字符串 "true"/"FaLse" 等会返回转换后的 bool 值

Why do you prefer literally true/false to empty?

**Comment by czpmango** [comment]
> > * 字符串 "true"/"FaLse" 等会返回转换后的 bool 值
> 
> Why do you prefer literally true/false to empty?

Consistent with the openCypher standard.

neo4j test:
```
neo4j@neo4j> return toBoolean("FaLse");
+--------------------+
| toBoolean("FaLse") |
+--------------------+
| FALSE              |
+--------------------+
```

**Review by panda-sheep — DISMISSED** [review]
Good job，lgtm

**Review by panda-sheep — APPROVED** [review]

**Review by panda-sheep — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**